### ISSUE TITLE
GetNumberOfLabelObjects() missing in BinaryImageToLabelMapFilter

### ISSUE THREAD

**Issue by yliu7366** [SimpleITK/SimpleITK#317]
Testing code (SimpleITK 1.0.1):

import matplotlib.pyplot as plt
%matplotlib inline
import SimpleITK as sitk

import numpy as np

from __future__ import print_function

print('processing...')

img = sitk.ReadImage('label.png')
thinImg = sitk.BinaryThinning(img)
labelImg = sitk.BinaryImageToLabelMap(thinImg, fullyConnected=True)

print(labelImg.GetPixelIDTypeAsString())
print(labelImg.GetNumberOfLabelObjects()) #AttributeError: 'Image' object has no attribute 'GetNumberOfLabelObjects'

**Comment by blowekamp** [comment]
I though I left a comment on the issue... Sorry if it was not conveyed.

The specific method you are trying to use was actually not working in ITK proper. Please see the [SimpleITK patch](#319) for more information.

The `RelabelComponentImageFilter` implements a similar algorithm. We have also [added additional methods there too](#303).

If you currently want to obtain the number of labels in an image before the above patches are merged, you can use the `RelabelComponentImageFilter` followed by the `StatisticsImageFilter` or the `ShapeLabelStatisticsImageFilter`.

**Comment by yliu7366** [comment]
Thank you Brad, I was able to use ConnectedComponent image filter followed by LabelShapeStatisticsImageFilter for my task. 

I also found out SimpleITK is missing LabelImageToShapeLabelMapFilter, and the LabelImageToLabelMapFilter in SimpleITK also missing the GetNumberOfLabelObjects() function. I feel these are all related and if it is necessary I can open a new ticket for it. Thanks again.

**Comment by blowekamp** [comment]
- Yes, the `sitk::ConnectedComponentsImageFilter` has a `GetObjectCount` method.

- The `itk::LabelImageToLabelMapFilter` filter does not have a "GetNumberOfLabelObjects" method, so that is not missing.
- The`sitk::ShapeLabelStatisticsImageFilter` does have a `GetNumberOfLabels` method. This class provides a interface similar to `sitk::LabelStatisticsImageFilter` for `itk::LabelImageToShapeLabelMapFilter`.

Why do you think they should have a `GetNumberOfLabelObjects` method?

**Comment by yliu7366** [comment]
I don't know, I was simply following ITK example to use SimpleITK. So here is the thing:

1. You have fixed the GetNumberOfLabelObjects() function for sitk::BinaryImageToLabelMap
2. I tested sitk::LabelImageToLabelMap and again the GetNumberOfLabelObjects() function is missing:

lll = sitk.LabelImageToLabelMap(labelImg)
print(lll.GetNumberOfLabelObjects())#AttributeError: 'Image' object has no attribute 'GetNumberOfLabelObjects'

To me the sitk::BinaryImageToLabelMap and the sitk::LabelImageToLabelMap are in the same category and that's why I said "I feel these are all related". If your fix to the sitk::BinaryImageToLabelMap will also update the sitk::LabelImageToLabelMap that will be great. 

Sorry my previous comment was not very accurate, instead saying "the LabelImageToLabelMapFilter in SimpleITK also missing the GetNumberOfLabelObjects() function", may be I should say comparing to ITK examples, the output of the sitk::LabelImageToLabelMapFilter doesn't have the GetNumberOfLabelObjects() function available and that may have been fixed by your patch to sitk::BinaryImageToLabelmap but I don't know.

**Comment by blowekamp** [comment]
What ITK example are you trying to follow? SimpleITK is a little different than ITK. Much of it is a straight forward translation, but SimpleITK's LabelMap abstraction is a little different.

In the following code:
```
lll = sitk.LabelImageToLabelMap(labelImg)
print(lll.GetNumberOfLabelObjects()) #AttributeError: 'Image' object has no attribute 'GetNumberOfLabelObjects'
```

`lll` is not a `sitk.LabelImageToLabelMapFilter` it is a `sitk::Image` object. You are calling the [procedural method for the `sitkLabelImageToLabelMapFilter`](https://itk.org/SimpleITKDoxygen/html/namespaceitk_1_1simple.html#a5900764d7069ea24b3829b3cd23998e4), which returns `sitk::Image`. The `sitk::Image` object does not have a `GetNumberOfLabelObjets`, by design unlike the `itk::LabelMap`.

 The SimpleITK `Image` class is a single template-less abstraction to the `itk::Image`, `itk::VectorImage` and [`itk::LabelMap`](https://itk.org/Doxygen/html/classitk_1_1LabelMap.html) classes and it can represent just about any pixel type too. More specifically, the SimpleITK's `LabelMap` representation is just templated over `itk::LabelMap<itk::LabelObject<PixelType,ImageDimension> >', and not over any of the label objects which contain statistics data. The SimpleITK design if for this statistics information to reside with the filter the compute them such as `sitk.ShapeLabelStatisticsImageFilter`. Additionally, most of the method for manipulating the label map are not in the `sitk::Image` class that operate beyond the basic `itk::Image` interface.

I hope that helps clarify the situation...but I unfortunately don't know what you are trying to do, so I can't tell you the SimpleITK way to do it.

**Comment by yliu7366** [comment]
It does help. I was trying to analysis label statistics and now I have found the proper pipeline in SimpleITK to use the ConnectedComponent followed by the LabelShapeStatisticsImageFilter.

It is in this learning procedure I realized the difference between SimpleITK and ITK. For example, this ITK [example](https://itk.org/Doxygen/html/WikiExamples_2ImageSegmentation_2LabelImageToLabelMapFilter_8cxx-example.html) shows the usage of GetNumberOfLabels() function on the output of the itk::LabelImageToLabelMapFilter. Now I understand, because of the design difference between SimpleITK and ITK, the usage is not directly transferable between them. 

So maybe a short description or notebook example explaining such difference (perhaps there's more difference between SimpleITK and ITK?)  is necessary to help novice users like me.
